doc/proxy-settings.md

# ProxySettings

Represents the proxy server configurations for API calls.

## Properties

| Name | Type | Tag | Description |
|  --- | --- | --- | --- |
| address | `String` | required | The proxy server URL. |
| port | `Integer` | optional | The port to connect to the proxy server. |
| username | `String` | optional | Username for proxy authentication. |
| password | `String` | optional | Password for proxy authentication. |

## Usage Example

```ruby
require 'shell_ev'
include ShellEv


client = ShellEv::Client.new(
  proxy_settings: ProxySettings.new(
    address: "http://localhost",
    port: 8888,
    username: 'user',
    password: 'pass'
  )
)
```



doc/environment-based-client-initialization.md

# Environment-Based Client Initialization

The SDK client can also be initialized directly from environment variables using the `from_env` class method. This allows the SDK to automatically read configuration values from the runtime environment or a `.env` file.

```ruby
require 'shell_ev'
include ShellEv

# Create client from environment
client = Client.from_env
```

You can also load values from a `.env` file before creating the client.
To do this, install and use the dotenv
gem:

```
gem install dotenv
```

Now require 'dotenv/load' automatically loads all variables from a `.env` file into `ENV`,
so the `from_env` method can access them.

```ruby
require 'dotenv/load'
require 'shell_ev'
include ShellEv

# Create client from environment
client = Client.from_env
```

The same method can accept keyword arguments to override any values read from the environment.

```ruby
# To override or fill in values that are not set in the environment, pass them
# as keyword arguments when calling `from_env`
client = Client.from_env(environment: 'sit', timeout: 30)
```

Values provided through arguments take precedence over those defined in environment variables.

## Example .env File

```ruby
ENVIRONMENT='sit'

O_AUTH_CLIENT_ID='oAuthClientId'
O_AUTH_CLIENT_SECRET='oAuthClientSecret'
O_AUTH_CLOCK_SKEW=5

TIMEOUT=60
MAX_RETRIES=3
BACKOFF_FACTOR=2
RETRY_INTERVAL=1
RETRY_STATUSES=408,413
RETRY_METHODS=GET,PUT,DELETE
PROXY_ADDRESS='http://localhost:3000'
PROXY_PORT=8080
PROXY_USERNAME='username'
PROXY_PASSWORD='password'
```

## Note

- If an environment variable is not defined, the default SDK configuration value will be used.



lib/shell_ev/configuration.rb
# shell_ev
#
# This file was automatically generated by APIMATIC v3.0 (
# https://www.apimatic.io ).

module ShellEv
  # An enum for SDK environments.
  class Environment
    ENVIRONMENT = [
      SIT = 'SIT'.freeze,
      PRODUCTION = 'Production'.freeze
    ].freeze

    # Converts a string or symbol into a valid Environment constant.
    def self.from_value(value, default_value = SIT)
      return default_value if value.nil?

      str = value.to_s.strip.downcase
      case str
      when 'sit' then SIT
      when 'production' then PRODUCTION

      else
        warn "[Environment] Unknown environment '#{value}', falling back to #{default_value} "
        default_value
      end
    end
  end

  # An enum for API servers.
  class Server
    SERVER = [
      OAUTH_SERVER = 'OAuth Server'.freeze,
      SHELL = 'Shell'.freeze
    ].freeze

    # Converts a string or symbol into a valid Server constant.
    def self.from_value(value, default_value = OAUTH_SERVER)
      return default_value if value.nil?

      str = value.to_s.strip.downcase
      case str
      when 'oauth_server' then OAUTH_SERVER
      when 'shell' then SHELL

      else
        warn "[Server] Unknown server '#{value}', falling back to #{default_value} "
        default_value
      end
    end
  end

  # All configuration including auth info and base URI for the API access
  # are configured in this class.
  class Configuration < CoreLibrary::HttpClientConfiguration
    def o_auth_client_id
      @client_credentials_auth_credentials.o_auth_client_id
    end

    def o_auth_client_secret
      @client_credentials_auth_credentials.o_auth_client_secret
    end

    def o_auth_token
      @client_credentials_auth_credentials.o_auth_token
    end

    # The attribute readers for properties.
    attr_reader :environment, :client_credentials_auth_credentials

    class << self
      attr_reader :environments
    end

    def initialize(
      connection: nil, adapter: :net_http_persistent, timeout: 60,
      max_retries: 0, retry_interval: 1, backoff_factor: 2,
      retry_statuses: [408, 413, 429, 500, 502, 503, 504, 521, 522, 524],
      retry_methods: %i[get put], http_callback: nil, proxy_settings: nil,
      environment: Environment::SIT, o_auth_client_id: nil,
      o_auth_client_secret: nil, o_auth_token: nil,
      client_credentials_auth_credentials: nil
    )
      super connection: connection, adapter: adapter, timeout: timeout,
            max_retries: max_retries, retry_interval: retry_interval,
            backoff_factor: backoff_factor, retry_statuses: retry_statuses,
            retry_methods: retry_methods, http_callback: http_callback,
            proxy_settings: proxy_settings

      # Current API environment
      @environment = String(environment)

      # OAuth 2 Client ID
      @o_auth_client_id = o_auth_client_id

      # OAuth 2 Client Secret
      @o_auth_client_secret = o_auth_client_secret

      # Object for storing information about the OAuth token
      @o_auth_token = if o_auth_token.is_a? OAuthToken
                        OAuthToken.from_hash o_auth_token.to_hash
                      else
                        o_auth_token
                      end

      # Initializing OAuth 2 Client Credentials Grant credentials with the provided auth parameters
      @client_credentials_auth_credentials = create_auth_credentials_object(
        o_auth_client_id, o_auth_client_secret, o_auth_token,
        client_credentials_auth_credentials
      )

      # The Http Client to use for making requests.
      set_http_client CoreLibrary::FaradayClient.new(self)
    end

    def clone_with(connection: nil, adapter: nil, timeout: nil,
                   max_retries: nil, retry_interval: nil, backoff_factor: nil,
                   retry_statuses: nil, retry_methods: nil, http_callback: nil,
                   proxy_settings: nil, environment: nil, o_auth_client_id: nil,
                   o_auth_client_secret: nil, o_auth_token: nil,
                   client_credentials_auth_credentials: nil)
      connection ||= self.connection
      adapter ||= self.adapter
      timeout ||= self.timeout
      max_retries ||= self.max_retries
      retry_interval ||= self.retry_interval
      backoff_factor ||= self.backoff_factor
      retry_statuses ||= self.retry_statuses
      retry_methods ||= self.retry_methods
      http_callback ||= self.http_callback
      proxy_settings ||= self.proxy_settings
      environment ||= self.environment
      client_credentials_auth_credentials = create_auth_credentials_object(
        o_auth_client_id, o_auth_client_secret, o_auth_token,
        client_credentials_auth_credentials || self.client_credentials_auth_credentials
      )

      Configuration.new(
        connection: connection, adapter: adapter, timeout: timeout,
        max_retries: max_retries, retry_interval: retry_interval,
        backoff_factor: backoff_factor, retry_statuses: retry_statuses,
        retry_methods: retry_methods, http_callback: http_callback,
        proxy_settings: proxy_settings, environment: environment,
        client_credentials_auth_credentials: client_credentials_auth_credentials
      )
    end

    def create_auth_credentials_object(o_auth_client_id, o_auth_client_secret,
                                       o_auth_token,
                                       client_credentials_auth_credentials)
      return client_credentials_auth_credentials if o_auth_client_id.nil? &&
                                                    o_auth_client_secret.nil? &&
                                                    o_auth_token.nil?

      warn('The \'o_auth_client_id\', \'o_auth_client_secret\', \'o_auth_token'\
           '\' params are deprecated. Use \'client_credentials_auth_credential'\
           's\' param instead.')

      unless client_credentials_auth_credentials.nil?
        return client_credentials_auth_credentials.clone_with(
          o_auth_client_id: o_auth_client_id,
          o_auth_client_secret: o_auth_client_secret,
          o_auth_token: o_auth_token
        )
      end

      ClientCredentialsAuthCredentials.new(
        o_auth_client_id: o_auth_client_id,
        o_auth_client_secret: o_auth_client_secret, o_auth_token: o_auth_token
      )
    end

    # All the environments the SDK can run in.
    ENVIRONMENTS = {
      Environment::SIT => {
        Server::OAUTH_SERVER => 'https://api-test.shell.com',
        Server::SHELL => 'https://api-test.shell.com/ev'
      },
      Environment::PRODUCTION => {
        Server::OAUTH_SERVER => 'https://api.shell.com',
        Server::SHELL => 'https://api.shell.com/ev'
      }
    }.freeze

    # Generates the appropriate base URI for the environment and the server.
    # @param [Configuration::Server] server The server enum for which the base URI is
    # required.
    # @return [String] The base URI.
    def get_base_uri(server = Server::SHELL)
      ENVIRONMENTS[environment][server].clone
    end

    # Builds a Configuration instance using environment variables.
    def self.build_default_config_from_env
      # === Core environment ===
      environment = Environment.from_value(ENV.fetch('ENVIRONMENT', 'sit'))
      timeout = (ENV['TIMEOUT'] || 60).to_f
      max_retries = (ENV['MAX_RETRIES'] || 0).to_i
      retry_interval = (ENV['RETRY_INTERVAL'] || 1).to_f
      backoff_factor = (ENV['BACKOFF_FACTOR'] || 2).to_f
      retry_statuses = ENV.fetch('RETRY_STATUSES',
                                 '[408, 413, 429, 500, 502, 503, 504, 521, 522, 524]').gsub(/[\[\]]/, '')
                                          .split(',')
                                          .map(&:strip)
                                          .map do |item|
                                            item.match?(/\A\d+\z/) ? item.to_i : item.downcase
                                          end
      retry_methods = ENV.fetch('RETRY_METHODS', '%i[get put]').gsub(/[\[\]]/, '')
                                          .split(',')
                                          .map(&:strip)
                                          .map do |item|
                                            item.match?(/\A\d+\z/) ? item.to_i : item.downcase
                                          end

      # === Authentication credentials ===
      client_credentials_auth_credentials = ClientCredentialsAuthCredentials.from_env

      # === Proxy settings ===
      proxy_settings = ProxySettings.from_env

      Configuration.new(
        environment: environment,
        timeout: timeout,
        max_retries: max_retries,
        retry_interval: retry_interval,
        backoff_factor: backoff_factor,
        retry_statuses: retry_statuses,
        retry_methods: retry_methods,
        client_credentials_auth_credentials: client_credentials_auth_credentials,
        proxy_settings: proxy_settings
      )
    end
  end
end
